pkg/util/provider/machinecontroller/machine_util.go

```diff
@@ -708,31 +718,78 @@ func (c *controller) reconcileMachineHealth(ctx context.Context, machine *v1alph
 					machine.Name,
 					timeOutDuration,
 				)
+				// Log the error message for machine failure
+				klog.Error(description)
+
+				clone.Status.LastOperation = v1alpha1.LastOperation{
+					Description:    description,
+					State:          v1alpha1.MachineStateFailed,
+					Type:           machine.Status.LastOperation.Type,
+					LastUpdateTime: metav1.Now(),
+				}
+				clone.Status.CurrentStatus = v1alpha1.CurrentStatus{
+					Phase: v1alpha1.MachineFailed,
+					//TimeoutActive:  false,
+					LastUpdateTime: metav1.Now(),
+				}
+				objectRequiresUpdate = true
 			} else {
-				// Timeour occurred due to machine being unhealthy for too long
+
+				// Timeout occurred due to machine being unhealthy for too long
 				description = fmt.Sprintf(
 					"Machine %s is not healthy since %s minutes. Changing status to failed. Node Conditions: %+v",
 					machine.Name,
 					timeOutDuration,
 					machine.Status.Conditions,
 				)
-			}
 
-			// Log the error message for machine failure
-			klog.Error(description)
+				machineSetName, machineDeployName := getMachineOwnerNames(machine)
+				//getting lock for machineDeployment this machine belongs to, creating lock if absent
+				tempCh := make(chan struct{}, 1)
+				val, loaded := c.safetyOptions.HealthChanMap.LoadOrStore(machineDeployName, tempCh)
+				if loaded {
+					close(tempCh)
+				} else {
+					klog.Infof("Lock for machineDeploy=%s created", machineDeployName)
+				}
+				ch := val.(chan struct{})
+
+				klog.Errorf("Number of goroutines now %d\n", runtime.NumGoroutine())
+				lockAcquireCtx, cancelFn := context.WithTimeout(ctx, lockAcquireTimeout)
+				defer cancelFn()
+
+				select {
+				case ch <- struct{}{}:
+					klog.Infof("Acquired lock ,machineName=%q , locks left to acquire %d", machine.Name, cap(ch)-len(ch))
```

Comment:
```suggestion
					klog.Infof("Acquired lock, machineName=%q", machine.Name)
```

Reply:
This log was introduced for debugging purposes, so I'll remove this at the end.